WalletWasabi/Models/SmartTransaction.cs

```diff
@@ -89,11 +91,40 @@ public SmartTransaction(Transaction transaction, Height height, uint256 blockHas
 
 		#endregion Constructors
 
-		public void SetHeight(Height height, uint256 blockHash = null, int blockIndex = 0)
+		/// <summary>
+		/// Update the transaction with the data acquired from another transaction. (For example merge their labels.)
+		/// </summary>
+		/// <param name="canUnconfirm">By default confirmed transaction won't get unconfirmed. You can override this option.</param>
+		public void Update(SmartTransaction tx, bool canUnconfirm = false)
 		{
-			Height = height;
-			BlockHash = blockHash;
-			BlockIndex = blockIndex;
+			// If this is not the same tx, then don't update.
+			if (this != tx)
+			{
+				throw new InvalidOperationException($"{GetHash()} != {tx.GetHash()}");
+			}
+
+			// Set the height related properties.
+			if (canUnconfirm)
+			{
+				Height = tx.Height;
+				BlockHash = tx.BlockHash;
+				BlockIndex = tx.BlockIndex;
+			}
+			else if (tx.Confirmed)
+			{
+				Height = tx.Height;
+				BlockHash = tx.BlockHash;
+				BlockIndex = tx.BlockIndex;
+			}
+
+			// Always the earlier seen is the firstSeen.
+			if (tx.FirstSeen < FirstSeen)
+			{
+				FirstSeen = tx.FirstSeen;
+			}
+
+			// Merge labels.
+			Label = SmartLabel.Merge(Label, tx.Label);
```

Comment:
I think that in most of the cases `Label` and `tx.Label` could be equal and the `Merge` method is not optimized for that situation. It could worth to check for equality in the `Merge` method or the `Label:ctor`.

Reply:
Done checking equality.